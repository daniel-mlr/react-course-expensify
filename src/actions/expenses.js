// actions/expenses.js
// import uuid from 'uuid'
import db from '../firebase/firebase'

// Étapes du traitement synchrone:
// ------------------------------
// components call action generator
// action generator returns object
// component dispatchs object
// redux stores changes

// Étapes du traitement asynchrone:
// --------------------------------
// components call action generator
// action generator returns a __function__
// component dispatch function (redux-thunk add support for that)
// function runs (has the ability to dispatch other actions and do whatever it wants)

// ADD_EXPENSE
//export const addExpense = ({description='', note='', amount=0, createdAt=0} = {}) => ({
export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
  // plus nécessaire
  // expense: {
  //   id: uuid(),
  //   description,
  //   note,
  //   amount,
  //   createdAt
  // }
})

export const startAddExpense = (expenseData = {}) => {
  // on retourne maintenant une fonction, plus un objet
  // cette fonction est appelée dans le code de redux avec comme argument 'dispatch'
  return (dispatch) => {
    const { description='', note='', amount=0, createdAt=0 } = expenseData
    const expense = {description, note, amount, createdAt}
    // retourne le résultat pour qu'on puisse le chainer dans les tests
    return db.ref('expenses').push(expense).then((ref) => {
      // ref est retourné si le push est réussi
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }))
    })
  }
}

// REMOVE_EXPENSE
export const removeExpense = ({ id }) => ({
  type: 'REMOVE_EXPENSE',
  id
})

export const startRemoveExpense = ({ id }) => {
  return (dispatch) => {
    return db.ref(`expenses/${id}`).remove().then(() =>{
      dispatch(removeExpense({ id }))
    })
  }
}

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
})

// SET_EXPENSES
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
})

export const startSetExpenses = () => {
  return (dispatch) => {
    const expenses = []

    return db.ref('expenses').once('value')
      .then((snapshot) => {
        snapshot.forEach((childSnapshot) => {
          expenses.push({
            id: childSnapshot.key,
            ...childSnapshot.val()
          })
        })
        dispatch(setExpenses(expenses))
      })
  }
}

