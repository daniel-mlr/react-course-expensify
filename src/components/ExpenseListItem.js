/* eslint react/prop-types: 0, quotes: 0 */
import React from 'react'
import { Link } from 'react-router-dom'
import moment from 'moment'
import numeral from 'numeral'

//export const ExpenseListItem = ({
const ExpenseListItem = ({
  id,
  // count,
  description,
  amount,
  createdAt
}) => (
  <Link className="list-item" to={'/edit/'.concat(id)}>
    <div>
      {/*<span>{count}</span>*/}
      <h3 className="list-item__title">{description}</h3>
      <span className="list-item__sub-title">
        {moment(createdAt).format('MMMM Do, YYYY')}
      </span>
    </div>
    <h3 className="list-item__data">{numeral(amount / 100).format('$0,0.00')}</h3>
  </Link> 
)
// plus besoin de dispatch. Donc:
export default ExpenseListItem
// utilisé dans la première version avec le bouton 'delete' adjoint à chaque
// item de la liste des expenses:
// export default connect()(ExpenseListItem)
//                ╰───────────────────────╯
//                            ▲
// ───────────────────────────┘
// simply give access to dispatch()
