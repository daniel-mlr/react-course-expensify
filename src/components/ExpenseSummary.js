// ExpenseSummary.js
/* eslint react/prop-types: 0 */

import React from 'react'
import {connect} from 'react-redux'
import selectExpenses from '../selectors/expenses'
import expensesTotal from '../selectors/expenses-total.js'


export const ExpenseSummary = (props) => (
  <p>Viewing {props.expenseCount} 
  expense{(props.expenseCount > 1 ) && 's'}&nbsp; 
  totalling {props.expenseTotal} </p> 
)

const mapStateToProps = (state) => {
  const expenses = selectExpenses(state.expenses, state.filters)
  return {
    expenseCount: expenses.length,
    expenseTotal: expensesTotal(expenses)
  }
}

export default connect(mapStateToProps)(ExpenseSummary)
