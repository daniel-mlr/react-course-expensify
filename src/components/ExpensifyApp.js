/* eslint react/prop-types: 0 */
import React, {Component} from 'react'

class ExpensifyApp extends Component {
  state = {
    options: [],
    selectedOption: undefined
  }

  handleCloseModal = () => {
    this.setState(() => ({selectedOption: undefined }))
  }
  handleAddOption = (option) => {
    // quelques validations
    if (!option) {
      return 'enter valid value to add item'
    } else if ( this.state.options.indexOf(option) > -1 ) {
      return 'this option already exists'
    }
    // ajout de la nouvelle option
    this.setState((prevState) => ({ options: prevState.options.concat(option)}))
  }

  handleDeleteOptions = () => this.setState( () => ({options: []}))

  handleDeleteOption = (optionToBeRemoved) => {
    this.setState((prevState) => ({
      options: prevState.options.filter((option) => optionToBeRemoved !== option )
    }))
  }

  handlePick = () => {
    const randomIx = Math.floor(Math.random() * this.state.options.length)
    // alert(this.state.options[randomIx])
    this.setState(() => ({selectedOption: this.state.options[randomIx]}))
  }

  /* life-cycle méthodes */
  componentDidMount() {
    // ('component est monté, aller chercher le data')
    try {
      const json = localStorage.getItem('options')
      const options = JSON.parse(json)
      if (options) { // au cas où json serait null (on veut '' qui est le défaut)
        this.setState(() => ({ options }))
      }
    } catch(e) {
      // do nothing
      // i.e. on utilise json seulement si le data est valide
    }

  }
  componentDidUpdate(prevProps, prevState) {
    // on vérifie si les données sont changées, car la sauvegarde
    // peut faire appel à une db externe, ce qui est très couteux.
    if (prevState.options.length !== this.state.options.length) {
      const json = JSON.stringify(this.state.options)
      localStorage.setItem('options', json)
    }
  }

  render() {
    return (
      <div>
        <div className="container">
        </div>
      </div>
    )
  }
}
export default ExpensifyApp
