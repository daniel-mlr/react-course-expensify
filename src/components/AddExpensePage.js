/* eslint react/prop-types: 0 */
import React from 'react'
import { connect } from 'react-redux'
import ExpenseForm from './ExpenseForm'
import { addExpense } from '../actions/expenses'

const  AddExpensePage = (props) => (
  <div>
    <h1>Add Expense</h1>
    <ExpenseForm 
      // passing the data out of the Expense Form
      // by calling a prop that get passed in by the parent
      // so that ExpenseForm can be reused for the edit page
      onSubmit={(expense) => {
        props.dispatch(addExpense(expense))
        props.history.push('/')
      }}
    />
  </div>
)
export default connect()(AddExpensePage)
