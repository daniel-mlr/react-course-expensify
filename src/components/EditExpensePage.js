/* eslint react/prop-types: 0 */
import React from 'react'
import { connect } from 'react-redux'
import ExpenseForm from './ExpenseForm'
import { startEditExpense, startRemoveExpense } from '../actions/expenses'

/*
const EditExpensePage = (props) => {
  console.log('dans edit expense page', props)
  return (
    <div>
      <ExpenseForm
        expense={props.expense}
        onSubmit={(expense) => {
          // props.dispatch(editExpense(props.match.params.id, expense))
          props.dispatch(editExpense(props.expense.id, expense))
          props.history.push('/')
        }}
      />
      <button onClick={() => {
        //removeExpense()
        props.dispatch(removeExpense({ id: props.expense.id }))
        props.history.push('/')
      }}>
      Remove
      </button>
    </div>
  )
}
*/

export class EditExpensePage extends React.Component {
  onSubmit = (expense) => {
    // this.props.editExpense(this.props.match.params.id, expense)
    this.props.editExpense(this.props.expense.id, expense)
    this.props.history.push('/')
  }
  onRemove = () => {
    this.props.removeExpense({ id: this.props.expense.id })
    this.props.history.push('/')
  }

  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h2 className="page-header__title">Edit Expense</h2>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit}/>
          <button className="button button--secondary" onClick={this.onRemove}>
            Remove Expense
          </button>
        </div>
      </div>
    )
  }
}

// give access to the current expense object
const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find(
      (expense) => expense.id === props.match.params.id
    )
  }
}
// selon instructions dans lecture no. 125 à 08:05 min.
// voir doc de react-redux pour mapDispatchToProps, le second argument est
// ownProps: il donne accès aux props passés au connected component
//const mapDispatchToProps = (dispatch, props) => ({
//     mais le nouvel argument *props* semble inutile ici.
const mapDispatchToProps = (dispatch) => ({
  editExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
  //removeExpense: ({id}) => dispatch(removeExpense( {id: id }))
  removeExpense: (data) => dispatch(startRemoveExpense(data))
})

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage)
