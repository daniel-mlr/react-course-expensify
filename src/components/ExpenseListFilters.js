/* eslint react/prop-types: 0 */
import React from 'react'
import {connect} from 'react-redux'
import { DateRangePicker } from 'react-dates'
import 'react-dates/lib/css/_datepicker.css'
import 'react-dates/initialize'
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters'

class ExpenseListFilters extends React.Component {
  onDatesChange = ({ startDate, endDate }) => {
    this.props.dispatch(setStartDate(startDate))
    this.props.dispatch(setEndDate(endDate))
  }
  onFocusChange = ( calenderFocused ) => {
    this.setState(() => ( { calenderFocused } ))
  }
  state = {
    calenderFocused: null
  }
  render() {
    return (
      <div>
        <input type="text" value={this.props.filters.text} onChange={(e) => {
          this.props.dispatch(setTextFilter(e.target.value))
        }}/>

        <select id="sel" value={this.props.filters.sortBy} onChange={(e) => {
          if (e.target.value==='date') {
            this.props.dispatch(sortByDate())
          } else {
            this.props.dispatch(sortByAmount())
          }
        }} >
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker
          startDate={this.props.filters.startDate}
          startDateId="dateDebutRecherche"
          endDate={this.props.filters.endDate}
          endDateId="dateFinRecherche"
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calenderFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
          showClearDates={true}
        />
      </div>
    )
  }
}


const mapStateToProps = (state) => {
  return state
}

export default connect(mapStateToProps)(ExpenseListFilters)

