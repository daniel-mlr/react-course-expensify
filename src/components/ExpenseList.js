/* eslint react/prop-types: 0 */
/* eslint jsx-quotes: ["error", "prefer-double"] */
import React from 'react'
import {connect} from 'react-redux'
import ExpenseListItem from './ExpenseListItem'
import selectExpenses from '../selectors/expenses'

// Regular unconnected component.
// Presentation component pattern.
// Get re-rendered with new prop values.
export const ExpenseList = (props) => (
  <div className="content-container">
    <div className="list-header">
      <div className="show-for-mobile">Expenses</div>
      <div className="show-for-desktop">Expense</div>
      <div className="show-for-desktop">Amount</div>
    </div>
    <div className="list-body">
      {
        props.expenses.length === 0 ? (
          <div className="list-item list-item--message">
            <span>No expenses yet.</span>
          </div>
        ) : (
          props.expenses.map((expense, index) => (
            <ExpenseListItem {...expense}
              key={expense.id}
              index={index}
              count={index + 1}
            />
          ))
        )
      }
    </div>
  </div>
)

// function that maps the store state to components props.
// Automatically rerun when state changes.
const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  }
}

// higher order component
export default connect(mapStateToProps)(ExpenseList)
//                     ╰─────────────╯  ╰─────────╯
//                           ▲                 ▲
//                           │                 │
// ce qu'on veux du 'store'──┘                 │
// component qu'on veut augmenter par le HOC ──┘
