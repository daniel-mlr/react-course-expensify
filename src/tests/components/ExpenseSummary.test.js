// ExpenseSummary.test.js

import React from 'react'
import { shallow } from 'enzyme'
import { ExpenseSummary } from '../../components/ExpenseSummary'
import expenses from '../fixtures/expenses'

test('should render ExpenseSummary with three expenses', () => {
  const wrapper = shallow(<ExpenseSummary 
    expenseCount={ expenses.length} 
    expenseTotal={ 89695 }
  />)
  expect(wrapper).toMatchSnapshot()
})

test('should render ExpenseSummary with 1 expense', () => {
  const wrapper = shallow(<ExpenseSummary 
    expenseCount={ [expenses[1]].length} 
    expenseTotal={ 85000 }
  />)
  expect(wrapper).toMatchSnapshot()
})
