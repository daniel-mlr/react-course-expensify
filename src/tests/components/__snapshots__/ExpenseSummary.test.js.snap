// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseSummary with 1 expense 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h2
      className="page-header__title"
    >
      Viewing 
      <span>
        1
      </span>
       
      expense
       totalling 
      <span>
        $850.00
      </span>
    </h2>
    <div
      className="page-header_actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render ExpenseSummary with three expenses 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h2
      className="page-header__title"
    >
      Viewing 
      <span>
        3
      </span>
       
      expenses
       totalling 
      <span>
        $896.95
      </span>
    </h2>
    <div
      className="page-header_actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;
